articles/actions.md

```diff
@@ -0,0 +1,363 @@
+- optionally monitoring the status for a goal from the action server
+- optionally requesting that the action server cancel an active goal
+- optionally checking the result for a goal received from the action server
+
+## Differences between ROS 1 and ROS 2 actions
+
+### First Class Support
+
+In ROS 1, actions are implemented as a separate library, [actionlib](http://wiki.ros.org/actionlib) that is built on top of the client libraries.
+This was done to avoid increasing the work required to create a client library in a new language, but actions turned out to be very important to packages like the [Navigation Stack](http://wiki.ros.org/navigation) and [MoveIt!](https://moveit.ros.org/)<sup>[1](#separatelib)</sup>.
+
+In ROS 2, actions will be included in the client library implementations.
+The work of writing a client library in a new language will be reduced by creating a common implementation in C.
+
+### Services used for Actions
+
+In ROS 1, actions were implemented using a set of topics under a namespace taken from the action name.
+ROS 1 services were not used because they are inherently synchronous and actions need to be asynchronous.
+Actions also needed to send status/feedback and be cancelable.
+
+In ROS 2, services are asynchronous in the common C implementation, so actions will use a combination of services and topics.
+
+### Goal Identifiers
+
+In ROS 1, an action client can create a goal ID when submitting a goal.
+This potentially leads to scenarios where mutliple clients independently generate the same goal ID.
+If an empty goal ID is sent, the action server will create one instead, which is not very useful since the client has no way to know the goal ID.
+
+In ROS 2, action clients will be the sole entities responsible for generating the goal ID.
+This way clients always know the goal ID for an action.
+Futhermore, a UUID will be used for each goal to mitigate the issue of goal ID collision across multiple clients.
+
+### Namespacing of Generated Messages and Services
+
+Multiple message and service definitions are generated from a single action definition.
+In ROS 1, the generated messages were prefixed with the name of the action to avoid conflicts with other messages and services.
+In ROS 2, the generated service and message definitions will exist in a different namespace to be impossible to conflict with non-action message and service definitions.
+For example, in Python the code from the generated definitions should be in the module `action` instead of `srv` and `msg`.
+In C++, the generated code should be in the namespace and folder `action` instead of `srv` and `msg`.
+
+### Visibility of Action Services and Topics
+
+In ROS 1, `rostopic list` would show all action topics in its output.
+In ROS 2, `ros2 topic list` and `ros2 service list` will not show topics and services used by actions by default.
+They can still be shown by passing an option to the commands to show hidden services and topics.
```

Comment:
The computational and memory burden of how to namespace action names hasn't been considered. Assuming that it is insignificant, it would be nice if the action names were consistent with topics and services: a node in namespace `/foo` with node name `bar` creates an action client/server with the name `~/baz` resulting in an action name `/foo/bar/baz`.

The topic and service name generation hasn't been decided. A name is hidden if any token starts with `_`. Inserting an `_action` token at the beginning or the end both seem like reasonable options. Thoughts on either of these options? 

#### After expansion to FQN topic and service names get `/_action/` prefix
When topics and service for action clients/servers are created the name must be expanded to a fully qualified name before creating publishers/clients/servers. Tools looking for actions only need to check if a topic or service name begins with `/_action/`. Humans looking at a list of topic or service names listed one-per-line left-justified would always see `/_action/` in the same place horizontally. Alphabetical sorting puts all action topic/service names together while the rest of the names are sorted by namespace.

* Action name: `~/move/something`
  * Topics
  ```
  /_action/name/space/nodename/move/something/status
  /_action/name/space/nodename/move/something/feedback
  ```
  * Services
  ```
  /_action/name/space/nodename/move/something/send_goal
  /_action/name/space/nodename/move/something/cancel
  /_action/name/space/nodename/move/something/get_result
  ```
* Action name: `move/something`
  * Topics
  ```
  /_action/name/space/move/something/status
  /_action/name/space/move/something/feedback
  ```
  * Services
  ```
  /_action/name/space/move/something/send_goal
  /_action/name/space/move/something/cancel
  /_action/name/space/move/something/get_result
  ```

* Action name: `/move/something`
  * Topics
  ```
  /_action/move/something/status
  /_action/move/something/feedback
  ```
  * Services
  ```
  /_action/move/something/send_goal
  /_action/move/something/cancel
  /_action/move/something/get_result
  ```

#### Topic and service names get `/_action` suffix
Code creating an action server doesn't need to expand an action name because `/_action` can always be appended to a name. Tools would need to read from the end of a string to tell if a service or topic name belongs to an action. It makes it harder for a human to visually spot actions since `_action` could be anywhere horizontally, but allows alphabetical sorting of topic names to group topic or service names by the intended namespace.

* Action name: `~/move/something`
  * Topics
  ```
  /name/space/nodename/move/something/status/_action
  /name/space/nodename/move/something/feedback/_action
  ```
  * Services
  ```
  /name/space/nodename/move/something/send_goal/_action
  /name/space/nodename/move/something/cancel/_action
  /name/space/nodename/move/something/get_result/_action
  ```
* Action name: `move/something`
  * Topics
  ```
  /name/space/move/something/status/_action
  /name/space/move/something/feedback/_action
  ```
  * Services
  ```
  /name/space/move/something/send_goal/_action
  /name/space/move/something/cancel/_action
  /name/space/move/something/get_result/_action
  ```

* Action name: `/move/something`
  * Topics
  ```
  /move/something/status/_action
  /move/something/feedback/_action
  ```
  * Services
  ```
  /move/something/send_goal/_action
  /move/something/cancel/_action
  /move/something/get_result/_action
  ```

Reply:
I prefer @wjwwood's proposal. I do not like the `/_action/` prefix at all. The `/_action` suffix is a distant second to @wjwwood 's proposal for me.

My reasoning is that I prefer topics and services to be grouped by namespace -> node -> type -> name, rather than grouping by type first or putting the type after the name. To me, this makes it easier to see related functionalities in topic/service/action lists, even when showing hidden topics.

Although @wjwwood's proposal has the type after the action name, it's a compromise I can accept to get the FQN action name to look like `/name/space/nodename/move/something` rather than always having to see `/name/space/nodename/move/_action/something` when using `ros2 action list`.

Reply:
Added @wjwwood's proposal in a4397401f83ebbb9dab22c7b04f35902909c950a

See also ros2/rcl#317

Reply:
Can the underlying topics and services used to formulate a action interfaces separated from normal ros topics and services? I think the bleedover between topics/services and actions will make security much more of a challenge.

E.g. how would user provision publish permission to all topics under /foo/bar/* , but simultaneously deny access to any actions under that same namespace? Requiring the careful use of filtering permissible topics/services based on a `_action` token seem rather fragel to me, and easily botched.

What if we made use of the the [ROS Specific Namespace Prefix](https://design.ros2.org/articles/topic_and_service_names.html#ros-specific-namespace-prefix) so we don't have to use '_action' tokens to denote the action interface from coliding namsepaced topics and services:


* Topics:
  * ra/bar/foo/status
  * ra/bar/foo/feedback
* Services:
  * ra/bar/foo/send_goal
  * ra/bar/foo/cancel
  * ra/bar/foo/get_result

Additionally, I'd prefer the `status`, `feedback`, `send_goal`, `cancel`, `get_result` get pushed down as tags, given they merely typing information and unecessarly collide with user namespacing. What happens if we want to extend this and add a new type, we'd have to worry about it colliding namesakes of existing community conventions.

E.g. what if a user had action namespaced under bar/foo/status, how would I differentiate access control permissions from /bar/foo/status and bar/foo/status/status using POSIX fnmatch rules?

Alternative to typing using tags:

* Topics:
  * dds_topic: {ra/bar/foo}
  * dds_tag: {name: ros2.action, value: status}
  * dds_tag: {name: ros2.action, value: feedback}
* Services:
  * dds_topic: {ra/bar/foo}
  * dds_tag: {name: ros2.action, value: send_goal}
  * dds_tag: {name: ros2.action, value: cancel}
  * dds_tag: {name: ros2.action, value: get_result}

Example permission hierarchy :

<details>

``` xml
<?xml version="1.0" encoding="utf-8"?>
<dds>
  <permissions>
    <grant name="action_caller">
      <subject_name>CN=action_caller</subject_name>
      <validity>
        <not_before>2013-10-26T00:00:00</not_before>
        <not_after>2023-10-26T22:45:30</not_after>
      </validity>
      <allow_rule>
        <domains>
          <id>0</id>
        </domains>
        <subscribe>
          <!-- subscribe to action topics  -->
          <topics>
            <topic>ra/bar/foo</topic>
          </topics>
          <data_tags>
            <tag>
              <name>ros2.action</name>
              <value>feedback</value>
              <value>status</value>
            </tag>
          </data_tags>
        </subscribe>
        <subscribe>
          <!-- subscribe to action replys  -->
          <topics>
            <topic>rr/bar/foo</topic>
          </topics>
          <data_tags>
            <tag>
              <name>ros2.action</name>
              <value>cancel</value>
              <value>get_result</value>
              <value>send_goal</value>
            </tag>
          </data_tags>
        </subscribe>
        <publish>
          <!-- publish to action requests  -->
          <topics>
            <topic>rq/bar/foo</topic>
          </topics>
          <data_tags>
            <tag>
              <name>ros2.action</name>
              <value>cancel</value>
              <value>get_result</value>
              <value>send_goal</value>
            </tag>
          </data_tags>
        </publish>
      </allow_rule>
      <default>DENY</default>
    </grant>
    <grant name="action_executor">
      <subject_name>CN=action_executor</subject_name>
      <validity>
        <not_before>2013-10-26T00:00:00</not_before>
        <not_after>2023-10-26T22:45:30</not_after>
      </validity>
      <allow_rule>
        <domains>
          <id>0</id>
        </domains>
        <publish>
          <!-- publish to action topics  -->
          <topics>
            <topic>ra/bar/foo</topic>
          </topics>
          <data_tags>
            <tag>
              <name>ros2.action</name>
              <value>feedback</value>
              <value>status</value>
            </tag>
          </data_tags>
        </publish>
        <publish>
          <!-- publish to action replys  -->
          <topics>
            <topic>rr/bar/foo</topic>
          </topics>
          <data_tags>
            <tag>
              <name>ros2.action</name>
              <value>cancel</value>
              <value>get_result</value>
              <value>send_goal</value>
            </tag>
          </data_tags>
        </publish>
        <subscribe>
          <!-- subscribe to action requests  -->
          <topics>
            <topic>rq/bar/foo</topic>
          </topics>
          <data_tags>
            <tag>
              <name>ros2.action</name>
              <value>cancel</value>
              <value>get_result</value>
              <value>send_goal</value>
            </tag>
          </data_tags>
        </subscribe>
      </allow_rule>
      <default>DENY</default>
    </grant>
  </permissions>
</dds>
```

Reply:
> E.g. how would user provision publish permission to all topics under /foo/bar/* , but simultaneously deny access to any actions under that same namespace? Requiring the careful use of filtering permissible topics/services based on a _action token seem rather fragel to me, and easily botched.

Isn't this something the security tools should be aware of? Like other tools, they should know that a topic name matching the action pattern is treated differently from a topic name matching the service pattern, for example. I'd go as far as to say that the namespace prefix concept (`/ra/bar` etc.) is exactly the same thing as using `_action` as a prefix.

> Additionally, I'd prefer the status, feedback, send_goal, cancel, get_result get pushed down as tags, given they merely typing information and unecessarly collide with user namespacing. What happens if we want to extend this and add a new type, we'd have to worry about it colliding namesakes of existing community conventions.

Without commenting on the idea of using tags, I do think that users cannot make topics that include defined system hidden keywords like `_action` and expect them to never collide. That's like using a variable name beginning with a double-underscore in C++ and assuming that it will never collide with a compiler's internal variables.

Reply:
> Isn't this something the security tools should be aware of? Like other tools, they should know that a topic name matching the action pattern is treated differently from a topic name matching the service pattern, for example. I'd go as far as to say that the namespace prefix concept (/ra/bar etc.) is exactly the same thing as using _action as a prefix.

It wouldn't matter if the security provisioning tools are made aware, given expressions are evaluated at the default DDS security plugin layer; the standard default plugin has no other notion of ROS2 other than what we can interject by aligning ROS2 subsystems with DDS primitives that serve as rule criterias: i.e. DDS Topics, Partitions, Data Tags, Domains.

If a user writes a ROS2 topic permission expression `/foo/bar/*` thinking it'll only apply to ROS2 topics (and so not actions), in the present case they would be in for a rude awakening when they find that parts of the action interface where also captured under that expression. In general, I think we'll find more users filter rules using fixed prefixes as opposed to fixed postfix. E.g `ra/*` seem a lot more sane to me than `rt/*_action`. E.g. it would be silly to attribute permission in a filesystem based on the extension types of the filenames as opposed to the folder hierarchy.

I think its ok if actions what to reuse ROS2 "topics" and "services" like interfaces, I'd only request that those meta "topics/services" specific to actions be sectioned off into its own root namespace reserved for actions. E.g If for some reason a user wanted to subscribe an meta "topic" of a given action namespace, they would ask the action library to create a topic subscriber, thus creating the datareader under `ra/.../status` rather than creating a datareader under `rt/.../status`.

> Without commenting on the idea of using tags, I do think that users cannot make topics that include defined system hidden keywords like _action and expect them to never collide. That's like using a variable name beginning with a double-underscore in C++ and assuming that it will never collide with a compiler's internal variables.

That challenge though is that it makes it hard to safely write generalised permission expressions for a particular ROS2 subsystem while not compromising another. If the two subsystems map into the same namespace, it becomes too unwieldy to seperate. Trying to separately secure a ROS2 network using topics, services and actions would then be like trying to solve the **Gordian Knot**! We must cut them apart!

Reply:
> What if we made use of the the ROS Specific Namespace Prefix so we don't have to use '_action' tokens to denote the action interface from coliding namsepaced topics and services:

The namespace prefix is specific to DDS middlewares. I would think if a middleware can choose to change a name by appending `rt/` then it could also choose to change a name containing `/_action/` to start with `ra/`. Time is short for Crystal. What's the minimum that could be done to make security rules easier without changing the `rmw` API?

Reply:
@ruffsl makes convincing arguments. I had forgotten that the security policies are applied well below the ROS layers.

Reply:
> The namespace prefix is specific to DDS middlewares. I would think if a middleware can choose to change a name by appending rt/ then it could also choose to change a name containing /_action/ to start with ra/.

@sloretz , are you here referring to the **"After expansion to FQN topic and service names get /_action/ prefix"** approach? I suppose having the middleware choose to change a name starting with `/_action/` to start with `ra/` instead isn't too bad. But I would question as to why we'd need to denote the action namespaces with `/_action/` to begin with? Aren't these namespace strings strongly typed above the middleware layer, e.g. how are namespace strings for service differentiated from namespace string for a topic like `rostopic://foo/bar` vs `rosservice://foo/bar` vs `rosaction://foo/bar`? If we want actions in ROS2 to be first class form of communication, I think they also deserve their own middleware mapping.

> Time is short for Crystal. What's the minimum that could be done to make security rules easier without changing the rmw API?

Well, my suggestion might be to map the meta topics and services to a seperate DDS topic namespaces using a different ROS Specific Namespace Prefix. The design specification allocated up to 8 characters we can use. If we are even going to use any more reserved chars for something cornerstone as actions, I'm not sure why we reserved so many for:

> While all planned prefixes consist of two characters, i.e. rX, anything proceeding the first namespace separator, i.e. /, can be considered part of the prefix. The standard reserves the right to use up to 8 characters for the prefix in case additional prefix space is needed in the future.
https://design.ros2.org/articles/topic_and_service_names.html

Here is an example of the patter I thought we'd have extrapolated to:

| ROS Subsystem        | Prefix |
|----------------------|--------|
| ROS Topics           | rt     |
| ROS Service Request  | rq     |
| ROS Service Response | rr     |
| ROS Service          | rs     |
| ROS Parameter        | rp     |
| **ROS Action Topics**          | **rat**     |
| **ROS Action Service Request**  | **raq**     |
| **ROS Action Service Response** | **rar**     |

Reply:
> are you here referring to the "After expansion to FQN topic and service names get /_action/ prefix" approach?

@ruffsl No, I was referring to the approach currently in the design doc where all topics/services have the form `<action name>/_action/...`. They all contain the text `/_action/`.

> But I would question as to why we'd need to denote the action namespaces with /_action/ to begin with? Aren't these namespace strings strongly typed above the middleware layer,

I'm not sure what you mean. As far as I know the URI scheme isn't sent to the middleware layer.

>  If we want actions in ROS2 to be first class form of communication, I think they also deserve their own middleware mapping.

That's a fair point. For context the rationale for not adding actions to `rmw` is here https://github.com/ros2/design/blob/a4397401f83ebbb9dab22c7b04f35902909c950a/articles/actions.md#actions-in-rmw.

> Well, my suggestion might be to map the meta topics and services to a seperate DDS topic namespaces using a different ROS Specific Namespace Prefix.

I think it's important to note that the `ROS Specific Namespace Prefix` is an agreement between the current DDS rmw implementations so they can talk to each other. It's not part of the `rmw` API. The most `rmw` exposes is an option to [ask a middleware to not change the name of a topic or service name](https://github.com/ros2/rmw/blob/c3b70771fd480a17a8d6bc5ae374dedb398a3c69/rmw/include/rmw/types.h#L208-L217).

That said, the aggreement can be extended to say if a topic or service name is being used by an action (determined by it containing the text `/_action/`) then change the prefix of the DDS topic. This wouldn't affect other middlewares or the `rmw` api.

@ruffsl If this route seems reasonable to you, may I ask you to open a pull request on the DDS topic mapping design doc? The table you have looks good to me with one comment that `ra` isn't necessary.

Reply:
Resolving as discussion has been moved to #203.